### ISSUE TITLE
[structural] Phase 2 = refactorisation fonctionnelle → JAX : l'étape manque, et la validation ne teste pas les gradients

### ISSUE THREAD

**Issue by maurinl26** [maurinl26/fortranspire#73]
## Définition retenue

**Phase 2 = Fortran → refactorisation fonctionnelle → JAX.** Chemin indépendant, **sans OpenACC** : ce n'est pas Phase 1 suivie d'une traduction, c'est une autre cible avec une autre étape intermédiaire.

L'étape qui porte tout est la **refactorisation fonctionnelle** — pureté, pas d'état global, pas de mutation, flux de données explicite. C'est ce qui rend `jit`, `grad` et `vmap` applicables. Sans elle, une traduction directe produit du code qui trace mais dont les transformations sont fausses.

## Ce que le pipeline fait aujourd'hui

`translation_graph.py` — **1 601 lignes monolithiques**, contre 81 pour Phase 1 qui délègue à `agent/nodes/`.

Graphe : `init → parser → explainer → dispatcher → translator → consolidator → …`

**Il n'y a aucune étape de fonctionnalisation.** Le LLM passe directement du Fortran impératif au JAX.

### Or les briques existent déjà — câblées uniquement sur le chemin OpenACC

| Nœud Phase 1 | Ce qu'il fait | Pourquoi c'est de la fonctionnalisation |
| --- | --- | --- |
| `nodes/extractor.py` | « Removes COMMON blocks and SAVE state by promoting them to explicit INTENT(INOUT) arguments » | **C'est la dé-globalisation** — la première moitié du travail |
| `nodes/pure_elemental.py` | Déterministe, **sans LLM** : règles sur `has_io` / `has_save` issues de Loki | `PURE` / `ELEMENTAL` est **le mot que Fortran emploie pour la pureté fonctionnelle** : pas d'effet de bord, pas d'état persistant |

> Un noyau que Loki refuse d'annoter `PURE` ne peut pas devenir une fonction JAX sans threading d'état explicite. C'est un **garde-fou déterministe et gratuit** — et Phase 2 ne s'en sert pas.

### 1. Aucune étape de fonctionnalisation

À ajouter, en réutilisant `extractor` et `pure_elemental`, plus ce qui est spécifique à JAX et n'existe nulle part :

- dépendances portées par les boucles → `lax.scan` / `fori_loop`
- mutation en place → mise à jour fonctionnelle `.at[].set()`
- branchement dépendant des données → `where` / `select_n` (un `if` sur un traced value casse `jit`)
- réductions → `jnp.sum` et consorts

Le plus possible de façon **déterministe, dans la couche AST** — pas confié au LLM.

### 2. La validation ne vérifie jamais la différentiabilité

`_validate_jax()` enchaîne : syntaxe → bytecode → `exec` → `make_jaxpr`. Cela prouve que **le code trace**. Cela ne prouve pas :

- l'**équivalence numérique** avec le Fortran d'origine — Phase 1 a `equivalence_harness`, Phase 2 n'a pas d'équivalent ;
- que les **gradients sont justes** — aucune comparaison `jax.grad` contre différences finies.

**C'est le manque le plus grave.** L'axe « utilité réelle » vient d'établir que tout l'intérêt de Phase 2 est la différentiabilité (ADflow, CATHARE, TGLF/QuaLiKiz, MODFLOW 6, FISPACT-II, CEA, DSSAT). Un noyau qui trace mais dont le gradient est faux est **pire qu'inutile** : il est silencieusement faux exactement là où le client s'y fie.

### 3. Aucune famille de prompts JAX

`prompts/` contient `extractor`, `openacc_kernel`, `openacc_driver`, `openmp_*`, `cython_*`, `doc_routine`. **Rien pour JAX.** Les prompts Phase 2 sont inline dans le monolithe — donc ni externalisés, ni versionnés, ni traduits, contrairement à ce que #3 a établi pour le reste.

## Périmètre

- Découper `translation_graph.py` en `agent/nodes_jax/` (ce que #2 a fait pour Phase 1 seulement).
- Nouveau graphe : `init → parser → extractor → functionalize → jax_kernel → equivalence → gradcheck`.
- Réutiliser `extractor` et `pure_elemental` **tels quels** — pas de fork.
- Nœud `functionalize` : déterministe partout où c'est possible, LLM en dernier recours.
- **`gradcheck`** : `jax.grad` contre différences finies sur le Fortran d'origine, tolérance déclarée. Bloquant.
- Étendre l'harnais d'équivalence de Phase 1 au couple Fortran ↔ JAX.
- Famille de prompts `prompts/jax_kernel/{en,fr}/v1.md`.
- Surfacer un **score de portabilité JAX** dans `analyze` et `explain`, comme le score GPU — un noyau non annotable `PURE` doit être signalé avant qu'on facture quoi que ce soit.

## Critères d'acceptation

- [ ] Phase 2 passe par une étape de fonctionnalisation explicite, traçable dans les sorties.
- [ ] Un noyau non `PURE`-able est refusé ou son état threadé explicitement — jamais ignoré.
- [ ] `gradcheck` bloquant, avec tolérance documentée.
- [ ] Équivalence numérique Fortran ↔ JAX vérifiée, pas seulement le tracing.
- [ ] Prompts JAX externalisés.
- [ ] `translation_graph.py` n'est plus un monolithe.

## Notes

**Priorité P1** : l'analyse d'utilité réelle a fait basculer sept cibles du panel vers la voie JAX, dont ADflow — où l'argument est de **supprimer 112 038 lignes d'adjoint généré par Tapenade**. On ne vend pas ça avec une validation qui s'arrête au tracing.

Réf : `translation_graph.py`, `agent/nodes/extractor.py`, `agent/nodes/pure_elemental.py`, #2 (split Phase 1), #3 (prompts externalisés), epic #65.

**Comment by maurinl26** [comment]
## Structure du reste de #73 — la différentiabilité par relaxation

Le premier lot est livré : `functionalize`, `gradcheck` bloquant, graphe Phase 2, prompts externalisés, `FORT030`. Il reste le sujet le plus intéressant, et il demande une distinction que l'issue initiale ne posait pas.

### La distinction qui structure tout

Rendre un kernel différentiable recouvre **deux opérations de nature différente**, et les confondre est exactement le mode d'échec silencieux qu'on cherche à éviter.

| | **Garde** | **Relaxation** |
| --- | --- | --- |
| Exemple | `sqrt(where(x>0, x, 1.0))` | `MAX(a,b)` → softmax β |
| Valeurs avant | **inchangées** | **modifiées** |
| Gradient | rendu fini | rendu continu |
| Nature | correction de traduction | **décision de modélisation** |
| Consentement | implicite | **doit être explicite** |

Une garde répare une traduction. **Une relaxation change le modèle physique.** Un `MAX` devenu softmax ne calcule plus la même chose — c'est parfois exactement ce qu'on veut (optimisation, UQ, adjoint), et c'est parfois inacceptable (un limiteur de flux qui doit rester TVD).

Conséquence directe : la relaxation ne peut pas être un défaut silencieux.

### Trois modes, explicites

```
--smoothing none      (défaut)  traduction fidèle ; gradients possiblement nuls ou indéfinis
--smoothing guarded             gardes seulement ; valeurs avant identiques
--smoothing smooth              relaxations ; valeurs avant modifiées, β/ε rapportés
```

### Catalogue des constructions non lisses

À détecter dans le Fortran, avec leur contrepartie. Colonne « limite » = ce vers quoi la forme relaxée converge.

| Construction Fortran | Problème | Relaxation | Limite |
| --- | --- | --- | --- |
| `MAX(a,b)` / `MIN(a,b)` | gradient en marche, nul d'un côté | `logaddexp(βa, βb)/β` | β→∞ ; biais `log 2/β` en `a=b` |
| `ABS(x)` | point anguleux en 0 | `sqrt(x²+ε²)` (pseudo-Huber) | ε→0 ; biais ε en 0 |
| `SIGN(x)` | échelon | `tanh(x/ε)` | ε→0 |
| `IF (x > t)` / `WHERE` | échelon | `σ((x−t)/ε)` en mélange | ε→0 |
| `MIN(MAX(x,lo),hi)` (limiteur) | deux points anguleux | composition softmin ∘ softmax | β→∞ |
| `FLOOR`, `NINT`, `INT` | localement plat, gradient nul | **aucune relaxation générale** — si c'est un index, c'est discret ; si c'est une table, interpoler | — |
| Table au plus proche voisin | gradient nul | interpolation linéaire / `jnp.interp` | exacte aux nœuds |
| `DO WHILE` de convergence | `lax.while_loop` **n'est pas différentiable en mode inverse** | nombre d'itérations fixe (`scan`), ou théorème des fonctions implicites | — |
| `SQRT(x)`, `x**0.5` en 0 | dérivée infinie | garde, pas relaxation | — |
| division par quantité annulable | dérivée infinie | garde | — |
| `MOD` / `MODULO` | dents de scie | forme périodique sin/cos si pertinent | — |
| tri, `MAXLOC`, `argmax` | discret | soft-argmax à température | β→∞ |

Deux entrées de ce tableau méritent d'être soulignées parce qu'elles ne sont pas des relaxations :

- **`FLOOR` / `NINT`** — c'est précisément la classe que `gradcheck` **ne sait pas détecter** (analytique et différence finie valent zéro toutes les deux). Une règle statique est donc le seul moyen de la voir. C'est l'argument le plus fort pour la détection AST.
- **`DO WHILE`** — `lax.while_loop` n'est pas différentiable en mode inverse. Un kernel de convergence traduit naïvement passera `gradcheck`… ou échouera à l'exécution du gradient. À traiter par un nombre d'itérations fixe ou par différentiation implicite.

### Ce qu'on n'écrit pas à la main

Les relaxations sont livrées comme **bibliothèque** (`fortranspire.jax_smooth`), pas générées par le LLM. Raisons :

- la stabilité numérique est facile à rater (`log(exp(βa)+exp(βb))` déborde ; `logaddexp` non) ;
- la limite ε→0 est **testable** une fois, pas à chaque génération ;
- le code émis importe des fonctions nommées, donc un relecteur voit immédiatement *où* le modèle a été relaxé et avec quel paramètre.

### Lots

1. **`fortranspire/jax_smooth.py`** — bibliothèque de gardes et de relaxations, chacune stable numériquement, avec test de convergence vers la forme dure.
2. **Détection `FORT031`** — construction non lisse repérée dans l'AST, nommée dans le message. Alimente `analyze` et `explain`.
3. **Mode `--smoothing`** dans le pipeline Phase 2 + le prompt, avec le catalogue.
4. **Rapport** — toute relaxation appliquée est listée dans la sortie, avec son paramètre. Non négociable : c'est ce qui rend la décision de modélisation visible.

### Ce qui reste hors de ce lot

- Équivalence numérique Fortran ↔ JAX (demande le build f2py de l'harnais Phase 1).
- Découpage du monolithe de 1 601 lignes — contourné plutôt que fait, le nouveau graphe ne passe plus par lui.
